papyri/gen.py

```diff
@@ -1050,6 +1052,99 @@ def _normalize_see_also(see_also: Section, qa: str):
     return new_see_also
 
 
+class PapyriDocTestRunner(doctest.DocTestRunner):
+    def __init__(self, *args, gen, obj, qa, config, **kwargs):
+        self.gen = gen
+        self.obj = obj
+        self.qa = qa
+        self.config = config
+        self._example_section_data = Section([], None)
+        super().__init__(*args, **kwargs)
+        import matplotlib
+        import matplotlib.pyplot as plt
+        import numpy as np
+
+        matplotlib.use("agg")
+
+        self.globs = {"np": np, "plt": plt, obj.__name__: obj}
+        self.globs.update(_get_implied_imports(obj))
+        for k, v in config.implied_imports.items():
+            self.globs[k] = obj_from_qualname(v)
+
+        self.figs = []
+
+    def _get_tok_entries(self, example):
+        entries = parse_script(
+            example.source, ns=self.globs, prev="", config=self.config, where=self.qa
+        )
+        if entries is None:
+            entries = [("jedi failed", "jedi failed")]
+        entries = _add_classes(entries)
+        tok_entries = [GenToken(*x) for x in entries]  # type: ignore
+        return tok_entries
+
+    def _figure_names(self):
+        """
+        File system can be case insensitive, we are not.
+        """
+        for i in count(0):
+            pat = f"fig-{self.qa}-{i}"
+            sha = sha256(pat.encode()).hexdigest()[:8]
+            yield f"{pat}-{sha}.png"
+
+    def report_start(self, out, test, example):
+        pass
+
+    def report_success(self, out, test, example, got):
+        import matplotlib.pyplot as plt
+
+        tok_entries = self._get_tok_entries(example)
+
+        self._example_section_data.append(
+            Code(tok_entries, got, ExecutionStatus.success)
+        )
+
+        figure_names = self._figure_names()
+
+        wait_for_show = self.config.wait_for_plt_show
+        fig_managers = _pylab_helpers.Gcf.get_all_fig_managers()
+        figs = []
+        if fig_managers and (("plt.show" in example.source) or not wait_for_show):
+            for fig, figname in zip(fig_managers, figure_names):
+                buf = io.BytesIO()
+                fig.canvas.figure.savefig(buf, dpi=300)  # , bbox_inches="tight"
+                buf.seek(0)
+                figs.append((figname, buf.read()))
+            plt.close("all")
+
+        for figname, _ in figs:
+            self._example_section_data.append(
+                Fig(
+                    RefInfo.from_untrusted(
+                        self.gen.root, self.gen.version, "assets", figname
+                    )
+                )
+            )
+        figs.extend(figs)
```

Comment:
```suggestion
        self.figs.extend(figs)
```

I think, or no figures will be saved I belive.